### ISSUE TITLE
feat: implement link-fragments rule (MD051)

### ISSUE THREAD

**Issue by shinagawa-web** [shinagawa-web/gomarklint#172]
## Overview

Implement the `link-fragments` rule (equivalent to markdownlint MD051 / remark-lint `no-undefined-references`).

Closes part of #76.

## Scope

gomarklint lints local Markdown files. Presets are limited to platforms with a local-file workflow (write locally → deploy/push). Browser-only authoring platforms are excluded even when their slug algorithm is documented.

## Problem

Internal anchor links (`#section`) are never validated by gomarklint. A broken fragment link silently compiles but leads to a 404 or missed scroll target at runtime.

markdownlint's MD051 hardcodes GitHub's slug algorithm, causing false positives on GitLab, Pandoc, kramdown, Zenn, Hugo, MkDocs, DocFX and others — each platform generates different slugs from the same heading text. This is the single most-reported class of MD051 issues in the markdownlint tracker.

## Proposed rule key

`link-fragments`

## Detection logic

1. Collect all headings in the document and compute their slugs using the configured algorithm.
2. For each inline link or reference link whose destination starts with `#`, check whether the fragment matches any computed slug.
3. Report a violation if no matching heading slug is found.

**Invalid:**
```markdown
## Introduction

See [Setup](#setup) for details.
```
(`#setup` does not match any heading slug)

**Valid:**
```markdown
## Introduction

See [Introduction](#introduction) for details.
```

### Preset (recommended)

```json
"link-fragments": {
  "enabled": true,
  "severity": "error",
  "slug-algorithm": "github"
}
```

### Custom (for platforms not covered by presets)

```json
"link-fragments": {
  "enabled": true,
  "slug-algorithm": "custom",
  "slug-params": {
    "lowercase": true,
    "preserve-unicode": true,
    "space-replacement": "-",
    "strip-chars": "[^\\w\\- ]",
    "collapse-separators": true
  }
}
```

| Parameter | Type | Description |
|---|---|---|
| `lowercase` | bool | Lowercase the entire heading before processing |
| `preserve-unicode` | bool | Keep non-ASCII characters (CJK, etc.) in the slug |
| `space-replacement` | string | Character to replace spaces (`"-"` or `"_"`) |
| `strip-chars` | regex | Characters to remove after replacement |
| `collapse-separators` | bool | Collapse consecutive separators into one |

Presets are sugar — each preset maps to a fixed set of `slug-params` internally.

### Built-in presets

| Preset | lowercase | preserve-unicode | space-replacement | strip-chars | collapse-separators | Notes |
|---|---|---|---|---|---|---|
| `github` | true | true | `-` | Unicode punctuation/symbol | false | github-slugger; consecutive spaces become consecutive hyphens |
| `gitlab` | true | true | `-` | `[^\p{L}\p{N}_-]` | true | goldmark slugify; collapses consecutive separators unlike GitHub |
| `zenn` | true | true | `-` | preserves all non-space chars | false | markdown-it-anchor default; anchors are percent-encoded in HTML |
| `qiita` | true | true | `-` | `[^\p{Word}\- ]` | false | `downcase.gsub(/[^\p{Word}\- ]/u, "").tr(" ", "-")`; consecutive hyphens preserved |
| `pandoc` | true | false | `-` | `[^a-zA-Z0-9_-]` | true | `auto_identifiers` extension; strips non-ASCII |
| `pandoc-gfm` | true | true | `-` | Unicode punctuation | false | `gfm_auto_identifiers` extension; equivalent to GitHub |
| `kramdown` | true | false | `-` | `[^a-zA-Z0-9 -]` | true | `header_ids` extension default |
| `hugo` | true | true | `-` | Unicode punctuation/symbol | false | `autoHeadingIDType: github` (default); equivalent to github-slugger |
| `mkdocs` | true | false | `-` | NFKD then ASCII-encode | true | Python-Markdown toc.py default; `uslugify` variant preserves Unicode |
| `docfx` | false | false | `-` | `[^a-zA-Z0-9-_.]` | true | Markdig AutoIdentifiers; does not lowercase |
| `vitepress` | true | partial | `-` | NFKD, strip combining chars, punctuation→`-` | true | Accented Latin normalized to ASCII; CJK preserved |
| `mdbook` | true | true | `-` | non-alphanumeric except `_` and `-` (Rust `is_alphanumeric()`) | false | CJK preserved via Unicode alphanumeric check |
| `gitea` | true | true | `-` | Unicode punctuation/symbol | false | goldmark-based; `user-content-` prefix added |
| `sphinx` | true | false | `-` | NFKD then ASCII then `[^a-z0-9]+`→`-` | true | Non-Latin-only headings fall back to `id1`, `id2`, etc. |
| `eleventy` | true | false | `-` | `@sindresorhus/slugify` (transliterate to approximate ASCII) | true | Used via `IdAttributePlugin` |
| `bitbucket` | true | true | `_` | unknown (needs verification) | unknown | `markdown-header-` prefix; spaces replaced with `_` |
| `azure-devops` | true | true | `-` | non-RFC-3986-unreserved chars percent-encoded | false | Unicode Zs category→`-` |
| `myst` | true | true | `-` | Unicode punctuation/symbol | false | MyST-Parser (Python/Sphinx); documented as GitHub-compatible |

### Aliases (same algorithm as an existing preset)

| Platform | Resolves to | Source |
|---|---|---|
| `docusaurus` | `github` | Uses github-slugger directly |
| `gatsby` | `github` | `gatsby-remark-autolink-headers` uses github-slugger |
| `astro` / `starlight` | `github` | Documented in Astro official docs |
| `nuxt-content` | `github` | Uses rehype-slug (github-slugger wrapper) |
| `forgejo` | `gitea` | Fork of Gitea; identical algorithm |
| `quarto` | `pandoc` | Uses `auto_identifiers` extension by default |

### Excluded platforms

| Platform | Reason |
|---|---|
| note | Heading IDs are UUID v4; not derived from heading text |
| WordPress (Gutenberg) | Rendered HTML does not include `id` attributes on headings |
| Redmine | Auto ID generation disabled (`header_ids: nil`) |
| Medium | Heading anchors are random hex IDs; not derived from heading text |
| Contentful | Headless CMS; ID generation is client-side and platform-agnostic |
| Obsidian | Uses raw heading text as fragment (`#Heading Text`); no slug generation |
| Notion | IDs contain random UUID suffix; not reproducible from heading text |
| Confluence | No standardized algorithm; varies by Markdown import mode |
| Typora | Algorithm depends on export target format |
| Speakerdeck | Slide platform; heading anchor links are not applicable |
| Hashnode | Source not confirmed |
| GitBook | Source not confirmed |
| Docswell | Slide platform; source not confirmed |
| Hexo | Strips CJK characters (Latin-only transliteration table) |
| forem (DEV.to) | No local file workflow; Markdown is written in a browser editor |
| hatena | No local file workflow; primary authoring is through the blog editor |
| ghost | No local file workflow; content is written via the Admin UI |
| bookstack | No local file workflow; primarily a browser-edited wiki (Git sync is optional) |

## Fix

No autofix. Report the line number of the broken link and the fragment value that could not be resolved.

## References

- [markdownlint MD051](https://github.com/DavidAnson/markdownlint/blob/main/doc/md051.md)
- [markdownlint issue tracker — slug algorithm complaints](https://github.com/DavidAnson/markdownlint/issues?q=MD051+slug)

**Comment by shinagawa-web** [comment]
## Implementation Plan

### Files to create
- `internal/rule/link_fragments.go` — `CheckLinkFragments` function + slug engine (preset resolver + custom params)
- `internal/rule/link_fragments_slug.go` — slug algorithm implementations (presets + parameterized engine)
- `internal/rule/link_fragments_test.go` — unit tests (table-driven)
- `internal/rule/link_fragments_slug_test.go` — slug algorithm unit tests per preset and custom params
- `e2e/fixtures/link_fragments_valid.md` — valid fixture
- `e2e/fixtures/link_fragments_violation.md` — violation fixture
- `e2e/fixtures/config-link-fragments.json` — e2e config

### Files to modify
- `internal/config/config.go` — register `link-fragments` in `DefaultConfigJSON` and `Default()`
- `internal/linter/linter.go` — call `CheckLinkFragments` in `collectErrors()`
- `e2e/e2e_test.go` — add e2e test cases
- `cmd/root_bench_test.go` — add headings and fragment links that exercise the scan path without violations

### Checklist
- [ ] Implement parameterized slug engine with the following params:
  - `lowercase` (bool)
  - `preserve-unicode` (bool)
  - `space-replacement` (string)
  - `strip-chars` (regex)
  - `collapse-separators` (bool)
- [ ] Implement built-in presets as fixed param sets: `github`, `gitlab`, `pandoc`, `pandoc-gfm`, `kramdown`, `zenn`, `hugo`, `mkdocs`, `docfx`, `qiita`, `vitepress`, `mdbook`, `gitea`, `sphinx`, `eleventy`, `bitbucket`, `azure-devops`, `myst`
- [ ] Implement aliases: `docusaurus` / `gatsby` / `astro` / `starlight` / `nuxt-content` → `github`; `forgejo` → `gitea`; `quarto` → `pandoc`
- [ ] Implement `CheckLinkFragments(filename, lines, offset, opts)`
  - Collect all headings and compute slugs with the resolved algorithm
  - Detect inline links `[text](#fragment)` and reference links where destination starts with `#`
  - Skip links inside fenced code blocks and inline code spans
  - Report violation if fragment does not match any heading slug
- [ ] Write unit tests covering:
  - Each built-in preset with ASCII and CJK headings
  - Custom slug-params combinations
  - Resolved and unresolved fragments
  - Links inside fenced code blocks and inline code spans (should be ignored)
  - Offset handling for frontmatter
  - Multiple violations across a document
- [ ] Register rule in config (`DefaultConfigJSON` + `Default()`) — disabled by default
- [ ] Integrate in linter `collectErrors()`
- [ ] Add e2e fixtures and test cases
- [ ] Verify `TestBenchmarkContentIsViolationFree` still passes
- [ ] Run `make test-all`
